lib/membrane_wav/serializer.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Membrane.WAV.Serializer do
+  @moduledoc """
+  Element responsible for raw audio serialization to WAV format.
+
+  Creates WAV header (its description can be found with `Membrane.WAV.Parser`) from received caps
+  and puts it before audio samples. The element assumes that audio is in PCM format. `File length`
+  and `data length` can be calculated only after processing all samples, so these values are
+  invalid (always set to 0).
+
+  The element has one option - `frames_per_buffer`. User can specify number of frames sent in one
+  buffer when demand unit on the output is `:buffers`. One frame contains `bits per sample` x
+  `number of channels` bits.
+  """
+  use Membrane.Filter
+
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+  alias Membrane.Caps.Audio.Raw.Format
+
+  @file_length 0
+  @data_length 0
+
+  @audio_format 1
+  @format_chunk_length 16
+
+  def_options frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :always,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(state) do
+    state = Map.put(state, :header_created, false)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(:input, caps, _context, state) do
+    buffer = %Buffer{payload: create_header(caps)}
+    state = Map.merge(state, %{header_created: true, caps: caps})
+
+    {{:ok, caps: {:output, caps}, buffer: {:output, buffer}, redemand: :output}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, _size, _unit, _context, %{header_created: false} = state) do
+    {:ok, state}
+  end
+
+  def handle_demand(:output, size, :bytes, _context, %{header_created: true} = state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  def handle_demand(
+        :output,
+        buffers_count,
+        :buffers,
+        _context,
+        %{header_created: true, frames_per_buffer: frames, caps: caps} = state
+      ) do
+    size = buffers_count * Caps.frames_to_bytes(frames, caps)
+
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(:input, buffer, _context, state) do
+    {{:ok, buffer: {:output, buffer}}, state}
+  end
+
+  defp create_header(%Caps{channels: channels, sample_rate: sample_rate, format: format}) do
+    {_signedness, bits_per_sample, _endianness} = Format.to_tuple(format)
+
+    data_transmission_rate = ceil(channels * sample_rate * bits_per_sample / 8)
+    block_alignment_unit = ceil(channels * bits_per_sample / 8)
```

Comment:
You can use integer division here (`div/2`)
```suggestion
    data_transmission_rate = div(channels * sample_rate * bits_per_sample, 8)
    block_alignment_unit = div(channels * bits_per_sample, 8)
```

Reply:
It is not the same as `ceil/1` and in WAV specification it is said to put `ceil` here (there are examples of `bits_per_sample` not divisible by 8 in the specification)

Reply:
OK, I've never seen sample size not divisible by 8, but OK, if the spec says so, ignore my comment